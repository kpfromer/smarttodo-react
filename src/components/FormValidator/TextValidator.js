import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { isArray, isPlainObject } from 'lodash';

import { ValidatorContext } from './ValidatorContext';
import TextField from '@material-ui/core/TextField';

const validateShape = PropTypes.shape({
  text: PropTypes.string.isRequired,
  validate: PropTypes.func.isRequired
}).isRequired;

export class TextValidator extends Component {

  static propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    validate: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.arrayOf(validateShape),
      validateShape
    ]).isRequired,
    validateOnBlur: PropTypes.bool,
    textFieldProps: PropTypes.object,
    name: PropTypes.string,
    updateFormInput: PropTypes.func,
    initializeInput: PropTypes.func,
    removeInput: PropTypes.func,
  };

  static defaultProps = {
    validateOnBlur: false
  };

  state = {
    error: false,
    errorText: ''
  };

  validateValue = value => {
    const { updateFormInput, validate, name } = this.props;

    let error = false;
    let errorText = '';

    const validatationObject = validateObj => {
      const isError = !validateObj.validate(value);
      if (isError) {
        error = true;
        errorText = validateObj.text;
      }
    };

    if (isArray(validate)) {
      for (const validateObj of validate) {
        validatationObject(validateObj);
        if (error) {
          break;
        }
      }
    } else if (isPlainObject(validate)) {
      validatationObject(validate);
    } else {
      error = !validate(value);
    }

    if (updateFormInput) {
      updateFormInput({
        name,
        value,
        error
      });
    }
    this.setState({ error, errorText });

    return { error };
  };

  componentDidMount() {
    const { initializeInput, name, value } = this.props;
    if (initializeInput) {
      initializeInput(name, value, this.validateValue);
    }
  }

  componentWillUnmount() {
    const { removeInput, name } = this.props;
    if (removeInput) {
      removeInput(name);
    }
  }

  componentDidUpdate(prevProps) {
    const { value: oldValue, name: oldName } = prevProps;
    const { value: newValue, name: newName, removeInput, initializeInput } = this.props;

    if (oldValue !== newValue) {
      this.validateValue(newValue);
    }

    if (oldName !== newName) {
      removeInput(oldName);
      initializeInput(newName);
    }
  }

  handleChange = event => {
    const { value } = event.target;
    const { error } = this.validateValue(value);
    this.props.onChange({ value, error });
  };

  handleBlur = event => {
    if (this.props.validateOnBlur) {
      this.validateValue(event.target.value);
    }
  };

  render() {
    const { value, textFieldProps } = this.props;
    const { error, errorText } = this.state;
    return (
      <TextField
        {...textFieldProps}
        onBlur={this.handleBlur}
        value={value}
        error={error}
        onChange={this.handleChange}
        helperText={errorText}
      />
    );
  }
}

export default props =>
  <ValidatorContext.Consumer>
    {
      ({ updateFormInput, initializeInput, removeInput }) =>
        <TextValidator
          {...props}
          updateFormInput={updateFormInput}
          initializeInput={initializeInput}
          removeInput={removeInput}
        />
    }
  </ValidatorContext.Consumer>
