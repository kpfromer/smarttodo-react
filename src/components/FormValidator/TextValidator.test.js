import React from "react";
import { TextValidator } from "./TextValidator";
import { mount } from 'enzyme';
import { SetupComponent } from "react-component-setup";
import TextField from '@material-ui/core/TextField';

const getComponentWithContext = context => {
  // mock out the context you're using in the component
  jest.doMock('./ValidatorContext', () => {
    return {
      ValidatorContext: {
        Consumer: (props) => props.children(context)
      }
    }
  });

  // you need to re-require after calling jest.doMock.
  // return the updated Component module that now includes the mocked context
  return require('./TextValidator').default;
};

describe('TextValidator', () => {
  let wrapper, mockHandleChange, mockValidate;

  beforeEach(() => {
    jest.resetModules();
    mockHandleChange = jest.fn();
    mockValidate = jest.fn();
  });

  describe('without context', () => {
    beforeEach(() => {
      const { shallow: setup } = SetupComponent({
        component: TextValidator,
        defaultProps: {
          value: 'initial',
          onChange: mockHandleChange,
          validate: mockValidate,
          textFieldProps: {
            miscOptionsToBeAddedToTextField: 'hello world'
          }
        }
      });

      ( { wrapper } = setup() );
    });

    it('renders a TextField with value and other properties', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('user enters text', () => {
      let newValue, valid, simulateChange, errorText;

      beforeEach(() => {
        newValue = 'new value';
        valid = false;

        simulateChange = () => {
          wrapper.find(TextField).props().onChange({ target: { value: newValue } });
          wrapper.update();
        };
      });
      // TODO: extract into library? pull request into jest-each?
      // simple version of a parameterized test, see jest-each
      // TODO: refactor even further, common setup - simulateChange, switch case based on type
      [
        {
          type: 'function',
          setup: () => {
            mockValidate.mockReturnValue(valid);
          }
        },
        {
          type: 'object',
          setup: () => {
            errorText = 'A ERROR! AHHHHHHHH!';
            wrapper.setProps({
              validate: {
                text: errorText,
                validate: mockValidate
              }
            });
          }
        },
        {
          type: 'array',
          setup: () => {
            errorText = 'SECOND ERROR!';
            wrapper.setProps({
              validate: [
                {
                  text: 'FIRST',
                  validate: () => true
                },
                {
                  text: errorText,
                  validate: mockValidate
                }
              ]
            });
          }
        }
      ].forEach(valueType => {
        const { type, setup } = valueType;
        describe(`when \`validate\` is a ${type}`, () => {
          beforeEach(() => {
            setup();
            simulateChange();
          });

          it('validates value', () => {
            expect(mockValidate).toHaveBeenCalledWith(newValue);
          });

          it('passes change and validity up to parent component', () => {
            expect(mockHandleChange).toHaveBeenCalledWith({
              error: !valid,
              value: newValue
            });
          });

          it('renders TextField with error if invalid', () => {
            expect(wrapper.find(TextField)).toHaveProp('error', true);
          });

          switch (type) {
            case 'object':
            case 'array':
              it('renders error helperText', () => {
                expect(wrapper.find(TextField)).toHaveProp('helperText', errorText);
              });
          }
        });
      });


      // describe('when `validate` is a function', () => {
      //   beforeEach(() => {
      //     mockValidate.mockReturnValue(valid);
      //     simulateChange();
      //   });
      //
      //   it('validates value', () => {
      //     expect(mockValidate).toHaveBeenCalledWith(newValue);
      //   });
      //
      //   it('passes change and validity up to parent component', () => {
      //     expect(mockHandleChange).toHaveBeenCalledWith({
      //       error: !valid,
      //       value: newValue
      //     });
      //   });
      //
      //   it('renders TextField with error if invalid', () => {
      //     expect(wrapper.find(TextField)).toHaveProp('error', true);
      //   });
      // });
      //
      // describe('when `validate` is an object', () => {
      //   let errorText;
      //   beforeEach(() => {
      //     errorText = 'A ERROR! AHHHHHHHH!';
      //     wrapper.setProps({
      //       validate: {
      //         text: errorText,
      //         validate: mockValidate
      //       }
      //     });
      //     simulateChange();
      //   });
      //
      //   it('validates value', () => {
      //     expect(mockValidate).toHaveBeenCalledWith(newValue);
      //   });
      //
      //   it('passes change and validity up to parent component', () => {
      //     expect(mockHandleChange).toHaveBeenCalledWith({
      //       error: !valid,
      //       value: newValue
      //     });
      //   });
      //
      //   it('renders TextField with error if invalid', () => {
      //     expect(wrapper.find(TextField)).toHaveProp('error', true);
      //   });
      //
      //   it('renders error helperText', () => {
      //     expect(wrapper.find(TextField)).toHaveProp('helperText', errorText);
      //   });
      // });
      //
      // describe('when `validate` is an array', () => {
      //   let errorText;
      //   beforeEach(() => {
      //     errorText = 'SECOND ERROR!';
      //     wrapper.setProps({
      //       validate: [
      //         {
      //           text: 'FIRST',
      //           validate: () => true
      //         },
      //         {
      //           text: errorText,
      //           validate: mockValidate
      //         }
      //       ]
      //     });
      //     simulateChange();
      //   });
      //
      //   it('validates value', () => {
      //     expect(mockValidate).toHaveBeenCalledWith(newValue);
      //   });
      //
      //   it('passes change and validity up to parent component', () => {
      //     expect(mockHandleChange).toHaveBeenCalledWith({
      //       error: !valid,
      //       value: newValue
      //     });
      //   });
      //
      //   it('renders TextField with error if invalid', () => {
      //     expect(wrapper.find(TextField)).toHaveProp('error', true);
      //   });
      //
      //   it('renders error helperText', () => {
      //     expect(wrapper.find(TextField)).toHaveProp('helperText', errorText);
      //   });
      // });
    });
  });

  describe('with context', () => {
    let name, mockUpdateFormInput, mockInitializeInput, mockRemoveInput, mockValidate;
    beforeEach(() => {
      name = 'usernameInput';

      mockUpdateFormInput = jest.fn();
      mockInitializeInput = jest.fn();
      mockRemoveInput = jest.fn();
      mockValidate = jest.fn();

      const Component = getComponentWithContext({
        updateFormInput: mockUpdateFormInput,
        initializeInput: mockInitializeInput,
        removeInput: mockRemoveInput
      });

      wrapper = mount(
        <Component
          name={name}
          value="initial"
          onChange={() => {
          }}
          validate={mockValidate}
        />
      );
    });
    // TODO: rename properties
    it('initializes input on creation', () => {
      expect(mockInitializeInput).toHaveBeenCalledWith(name);
    });
    it('deinitializes and reinitializes if name changes', () => {
      const differentName = 'adifferentname';
      wrapper.setProps({
        name: differentName
      });
      expect(mockRemoveInput).toHaveBeenCalledWith(name);
      expect(mockInitializeInput).toHaveBeenCalledWith(differentName);
    });
    it('updates form input error on TextField value change', () => {
      const invalid = true;

      mockValidate.mockReturnValue(!invalid);
      wrapper.find('TextField').props().onChange({ target: { value: 'new invalid value' } });
      wrapper.update();

      expect(mockUpdateFormInput).toHaveBeenCalledWith({
        name,
        error: invalid
      })
    });
    it('deinitializes input on unmount', () => {
      wrapper.unmount();
      expect(mockRemoveInput).toHaveBeenCalledWith(name);
    });
  });
});