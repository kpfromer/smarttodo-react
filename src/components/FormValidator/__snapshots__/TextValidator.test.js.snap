// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextValidator without context renders a TextField with value and other properties 1`] = `
<TextField
  error={false}
  helperText=""
  miscOptionsToBeAddedToTextField="hello world"
  onBlur={[Function]}
  onChange={[Function]}
  required={false}
  select={false}
  value="initial"
  variant="standard"
/>
`;
